src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs

```diff
@@ -19,52 +19,70 @@ internal static class SegmentedArrayHelper
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal static int GetSegmentSize<T>()
         {
+            return Unsafe.SizeOf<T>() switch
+            {
+                // Hard code common values since not all versions of .NET support constant-folding the calculation.
+                // Values are validated against the reference implementation in CalculateSegmentSize in unit tests.
+                4 => 16384,
+                8 => 8192,
+                12 => 4096,
+                16 => 4096,
+                24 => 2048,
+                28 => 2048,
+                32 => 2048,
+                40 => 2048,
 #if NETCOREAPP3_0_OR_NEWER
-            return InlineCalculateSegmentSize(Unsafe.SizeOf<T>());
+                _ => InlineCalculateSegmentSize(Unsafe.SizeOf<T>()),
 #else
-            if (Unsafe.SizeOf<T>() == Unsafe.SizeOf<object>())
-            {
-                return ReferenceTypeSegmentHelper.SegmentSize;
-            }
-            else
-            {
-                return ValueTypeSegmentHelper<T>.SegmentSize;
-            }
+                _ => ValueTypeSegmentHelper<T>.SegmentSize,
 #endif
+            };
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         internal static int GetSegmentShift<T>()
         {
+            return Unsafe.SizeOf<T>() switch
+            {
+                // Hard code common values since not all versions of .NET support constant-folding the calculation.
+                // Values are validated against the reference implementation in CalculateSegmentShift in unit tests.
+                4 => 14,
+                8 => 13,
+                12 => 12,
+                16 => 12,
+                24 => 11,
+                28 => 11,
+                32 => 11,
+                40 => 11,
 #if NETCOREAPP3_0_OR_NEWER
-            return InlineCalculateSegmentShift(Unsafe.SizeOf<T>());
+                _ => InlineCalculateSegmentShift(Unsafe.SizeOf<T>()),
 #else
-            if (Unsafe.SizeOf<T>() == Unsafe.SizeOf<object>())
```

Comment:
>if (Unsafe.SizeOf<T>() == Unsafe.SizeOf<object>()) [](http://example.com/codeflow?start=12&length=50)

Same suggestion about an assert.

Reply:
➡️ Addressing as part of https://github.com/dotnet/roslyn/pull/67921#discussion_r1175772691